File: app/src/main/java/hpn332/cb/utils/database/DBContract.java
package hpn332.cb.utils.database;

import android.net.Uri;

class DBContract {

	interface ProjectColumns {
		String ID          = "_id";
		String TITLE       = "title";
		String DESCRIPTION = "description";
	}

	interface TagColumns {
		String ID          = "_id";
		String TITLE       = "title";
		String DESCRIPTION = "description";
		String COLOR       = "color";
	}

	interface BackLogColumns {
		String ID          = "_id";
		String TITLE       = "title";
		String DESCRIPTION = "description";
		String PROJECT     = "project";
		String COLOR       = "color";
	}

	interface TaskColumns {
		String ID          = "_id";
		String TITLE       = "title";
		String DESCRIPTION = "description";
		String PROJECT     = "project";
		String BACKLOG     = "backlog";
		String TAGS        = "tags";
		String STEP        = "step";
		String RANK        = "rank";
	}

	static final String AUTHORITY     = "hpn332.cb";
	static final Uri    BASE_URI      = Uri.parse("content://" + AUTHORITY);
	static final String TABLE_PROJECT = "project";
	static final String TABLE_TASK    = "task";
	static final String TABLE_TAG     = "tag";
	static final String TABLE_BACKLOG = "backlog";

	static final Uri URI_PROJECT = BASE_URI.buildUpon()
			.appendEncodedPath(TABLE_PROJECT).build();

	static final Uri URI_TASK = BASE_URI.buildUpon()
			.appendEncodedPath(TABLE_TASK).build();

	static final Uri URI_TASK_STEP = BASE_URI.buildUpon()
			.appendEncodedPath(TABLE_TASK + "step").build();

	static final Uri URI_TAG = BASE_URI.buildUpon()
			.appendEncodedPath(TABLE_TAG).build();

	static final Uri URI_BACKLOG = BASE_URI.buildUpon()
			.appendEncodedPath(TABLE_BACKLOG).build();

	static class ProjectEntry implements ProjectColumns {

		static final String CREATE_TABLE =
				"CREATE TABLE " + TABLE_PROJECT + " ("
						+ ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
						+ TITLE + " TEXT,"
						+ DESCRIPTION + " TEXT"
						+ ");";

		static final String CONTENT_TYPE      =
				"vnd.android.cursor.dir/vnd." + AUTHORITY + ".project";
		static final String CONTENT_ITEM_TYPE =
				"vnd.android.cursor.item/vnd." + AUTHORITY + ".project";
	}

	static class TaskEntry implements TaskColumns {

		static final String CREATE_TABLE =
				"CREATE TABLE " + TABLE_TASK + " ("
						+ ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
						+ TITLE + " TEXT,"
						+ DESCRIPTION + " TEXT,"
						+ PROJECT + " INTEGER,"
						+ BACKLOG + " INTEGER,"
						+ TAGS + " TEXT,"
						+ STEP + " INTEGER,"
						+ RANK + " INTEGER"
						+ ");";

		static final String CONTENT_TYPE      =
				"vnd.android.cursor.dir/vnd." + AUTHORITY + ".task";
		static final String CONTENT_ITEM_STEP =
				"vnd.android.cursor.step.dir/vnd." + AUTHORITY + ".task";
		static final String CONTENT_ITEM_TYPE =
				"vnd.android.cursor.item/vnd." + AUTHORITY + ".task";
	}

	static class TagEntry implements TagColumns {

		static final String CREATE_TABLE =
				"CREATE TABLE " + TABLE_TAG + " ("
						+ ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
						+ TITLE + " TEXT,"
						+ DESCRIPTION + " TEXT,"
						+ COLOR + "  INTEGER"
						+ ");";

		static final String CONTENT_TYPE      =
				"vnd.android.cursor.dir/vnd." + AUTHORITY + ".tag";
		static final String CONTENT_ITEM_TYPE =
				"vnd.android.cursor.item/vnd." + AUTHORITY + ".tag";
	}

	static class BackLogEntry implements BackLogColumns {

		static final String CREATE_TABLE =
				"CREATE TABLE " + TABLE_BACKLOG + " ("
						+ ID + " INTEGER PRIMARY KEY AUTOINCREMENT,"
						+ TITLE + " TEXT,"
						+ DESCRIPTION + " TEXT,"
						+ PROJECT + " INTEGER,"
						+ COLOR + "  INTEGER"
						+ ");";

		static final String CONTENT_TYPE      =
				"vnd.android.cursor.dir/vnd." + AUTHORITY + ".backlog";
		static final String CONTENT_ITEM_TYPE =
				"vnd.android.cursor.item/vnd." + AUTHORITY + ".backlog";
	}


	static Uri buildUri(Uri URI, String id) {return URI.buildUpon().appendEncodedPath(id).build();}

	static String getPathSegments(Uri uri)  {return uri.getPathSegments().get(1);}
}

File: app/src/main/java/hpn332/cb/model/data/DBContract.kt
package hpn332.cb.model.data

import android.net.Uri

@Deprecated("Not use")
object DBContract {

    internal val AUTHORITY = "hpn332.cb"
    internal val BASE_URI = Uri.parse("content://" + AUTHORITY)
    internal val TABLE_PROJECT = "project"
    internal val TABLE_TASK = "task"
    internal val TABLE_TAG = "tag"
    internal val TABLE_BACKLOG = "backlog"

    val URI_PROJECT = buildUri(BASE_URI, TABLE_PROJECT)

    val URI_TASK = buildUri(BASE_URI, TABLE_TASK)

    val URI_TASK_STEP = buildUri(BASE_URI, TABLE_TASK + "step")

    val URI_TAG = buildUri(BASE_URI, TABLE_TAG)

    val URI_BACKLOG = buildUri(BASE_URI, TABLE_BACKLOG)

    class ProjectEntry {
        companion object {

            val ID = "id"
            val TITLE = "title"
            val DESCRIPTION = "description"


            internal val CONTENT_TYPE =
                "vnd.android.cursor.dir/vnd.$AUTHORITY.project_id"
            internal val CONTENT_ITEM_TYPE =
                "vnd.android.cursor.item/vnd.$AUTHORITY.project_id"
        }
    }

    class TaskEntry {
        companion object {

            val ID = "id"
            val TITLE = "title"
            val DESCRIPTION = "description"
            val PROJECT = "project_id"
            val BACKLOG = "backlog"
            val TAGS = "tags"
            val STEP = "step"
            val RANK = "rank"


            internal val CONTENT_TYPE =
                "vnd.android.cursor.dir/vnd.$AUTHORITY.task"
            internal val CONTENT_ITEM_STEP =
                "vnd.android.cursor.step.dir/vnd.$AUTHORITY.task"
            internal val CONTENT_ITEM_TYPE =
                "vnd.android.cursor.item/vnd.$AUTHORITY.task"
        }
    }

    class TagEntry {
        companion object {

            val ID = "id"
            val TITLE = "title"
            val DESCRIPTION = "description"
            val COLOR = "color"


            internal val CONTENT_TYPE =
                "vnd.android.cursor.dir/vnd.$AUTHORITY.tag_id"
            internal val CONTENT_ITEM_TYPE =
                "vnd.android.cursor.item/vnd.$AUTHORITY.tag_id"
        }
    }

    class BackLogEntry {
        companion object {

            val ID = "id"
            val TITLE = "title"
            val DESCRIPTION = "description"
            val PROJECT = "project_id"
            val COLOR = "color"


            internal val CONTENT_TYPE =
                "vnd.android.cursor.dir/vnd.$AUTHORITY.backlog"
            internal val CONTENT_ITEM_TYPE =
                "vnd.android.cursor.item/vnd.$AUTHORITY.backlog"
        }
    }


    fun buildUri(URI: Uri, id: String): Uri = URI.buildUpon().appendEncodedPath(id).build()

    internal fun getPathSegments(uri: Uri): String = uri.pathSegments[1]
}
